### PR TITLE
test: check that combining PSBTs with different txs fails

### PR DESCRIPTION
**Pull request by theStack**
This PR adds missing test coverage for the `combinepsbt` RPC, in the case of combining two PSBTs with different transactions:
https://github.com/bitcoin/bitcoin/blob/b8067cd435059eedb580975afc62c4e7a6f27321/src/psbt.cpp#L24-L27
The calling function `CombinePSBTs` checks for the false return value and then returns the transaction error string `PSBT_MISMATCH`:
https://github.com/bitcoin/bitcoin/blob/b8067cd435059eedb580975afc62c4e7a6f27321/src/psbt.cpp#L433-L435
https://github.com/bitcoin/bitcoin/blob/b8067cd435059eedb580975afc62c4e7a6f27321/src/util/error.cpp#L30-L31

### CODE REVIEW

**Comment by theStack** [comment]
Force-pushed a different variant of the first commit (extending the PSBT ctor), avoiding mutable objects as default parameters due to the following linter complaint:
```
A mutable list or dict seems to be used as default parameter value:

test/functional/test_framework/psbt.py:    def __init__(self, g=PSBTMap(), i=[], o=[]):

This is how mutable list and dict default parameter values behave:

>>> def f(i, j=[], k={}):
...     j.append(i)
...     k[i] = True
...     return j, k
...
>>> f(1)
([1], {1: True})
>>> f(1)
([1, 1], {1: True})
>>> f(2)
([1, 1, 2], {1: True, 2: True})

The intended behaviour was likely:

>>> def f(i, j=None, k=None):
...     if j is None:
...         j = []
...     if k is None:
...         k = {}
...     j.append(i)
...     k[i] = True
...     return j, k
...
>>> f(1)
([1], {1: True})
>>> f(1)
([1], {1: True})
>>> f(2)
([2], {2: True})
```

File: test/functional/rpc_psbt.py [hunk 1]
```diff
@@ -820,6 +820,15 @@ def test_psbt_input_keys(psbt_input, keys):
...
             assert_equal(res_input[preimage_key][hash.hex()], preimage.hex())
 
+        self.log.info("Test that combining PSBTs with different transactions fails")
+        tx = CTransaction()
+        tx.vin = [CTxIn(outpoint=COutPoint(hash=int('aa' * 32, 16), n=0), scriptSig=b"")]
+        tx.vout = [CTxOut(nValue=0, scriptPubKey=b"")]
+        psbt1 = PSBT(PSBTMap({PSBT_GLOBAL_UNSIGNED_TX: tx.serialize()}), [PSBTMap()], [PSBTMap()]).to_base64()
+        tx.vout[0].nValue += 1  # slightly modify tx
+        psbt2 = PSBT(PSBTMap({PSBT_GLOBAL_UNSIGNED_TX: tx.serialize()}), [PSBTMap()], [PSBTMap()]).to_base64()
+        assert_raises_rpc_error(-8, "PSBTs not compatible (different transactions)", self.nodes[0].combinepsbt, [psbt1, psbt2])
```
Comment (instagibbs) [inline test/functional/rpc_psbt.py:830]: Might as well check the naive `[psbt1, psbt1]` case to make sure that works

File: test/functional/rpc_psbt.py [hunk 2]
```diff
@@ -820,6 +820,15 @@ def test_psbt_input_keys(psbt_input, keys):
             assert hash.hex() in res_input[preimage_key]
             assert_equal(res_input[preimage_key][hash.hex()], preimage.hex())
 
+        self.log.info("Test that combining PSBTs with different transactions fails")
+        tx = CTransaction()
+        tx.vin = [CTxIn(outpoint=COutPoint(hash=int('aa' * 32, 16), n=0), scriptSig=b"")]
+        tx.vout = [CTxOut(nValue=0, scriptPubKey=b"")]
+        psbt1 = PSBT(PSBTMap({PSBT_GLOBAL_UNSIGNED_TX: tx.serialize()}), [PSBTMap()], [PSBTMap()]).to_base64()
```
Comment (instagibbs) [inline test/functional/rpc_psbt.py:827]: could use named args here

File: test/functional/rpc_psbt.py [hunk 3]
```diff
@@ -820,6 +820,15 @@ def test_psbt_input_keys(psbt_input, keys):
...
+        tx.vout[0].nValue += 1  # slightly modify tx
+        psbt2 = PSBT(PSBTMap({PSBT_GLOBAL_UNSIGNED_TX: tx.serialize()}), [PSBTMap()], [PSBTMap()]).to_base64()
```
Comment (instagibbs) [inline test/functional/rpc_psbt.py:829]: could use named args here

**Review by instagibbs — APPROVED** [review]
ACK b4f332275c593d5c8f454f3e2b7f5d491ef4f520

File: test/functional/rpc_psbt.py (hunk 1 above)
Comment (theStack) [reply test/functional/rpc_psbt.py:830]: Good idea, done.

File: test/functional/rpc_psbt.py (hunk 2 above)
Comment (theStack) [reply test/functional/rpc_psbt.py]: Thanks, done, also changed the PSBT ctor to enforce named arguments.

**Comment by theStack** [comment]
Force-pushed with changes as suggested by @instagibbs (named arguments for PSBT ctor, adding call for success case of combining two identical PSBTs).

**Comment by instagibbs** [comment]
reACK https://github.com/bitcoin/bitcoin/pull/25670/commits/4e616d20c9e92b5118a07d4a4b8562fffc66e767

**Comment by achow101** [comment]
ACK 4e616d20c9e92b5118a07d4a4b8562fffc66e767

**Pull request merged**